### PR TITLE
Scheduler can recieve its policy configuration from a ConfigMap

### PR DESCRIPTION
**Pull request by bsalamat**
**What this PR does / why we need it**: This PR adds the ability to scheduler to receive its policy configuration from a ConfigMap. Before this, scheduler could receive its policy config only from a file. The logic to watch the ConfigMap object will be added in a subsequent PR.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Release note**:

```Add the ability to the default scheduler to receive its policy configuration from a ConfigMap object.
```

### CODE REVIEW

**Comment by bsalamat** [comment]
@kubernetes/sig-scheduling-pr-reviews

**Comment by bsalamat** [comment]
@davidopp

**Comment by bsalamat** [comment]
Looks like I need to manually add some conversion functions. I will add them soon.

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (gyliu513) [inline plugin/cmd/kube-scheduler/app/options/options.go]: s/It must exist in the system namespace before scheduler initialization./It must exist in the system namespace before scheduler initialization if scheduler is not using legacy policy config.

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (gyliu513) [inline plugin/cmd/kube-scheduler/app/options/options.go]: As the `PolicyConfigMapName` and `PolicyConfigFile` are some internal used words, how about s/PolicyConfigMapName and uses PolicyConfigFile/policy config map and uses policy config file.

File: pkg/apis/componentconfig/types.go
Comment (timothysc) [inline pkg/apis/componentconfig/types.go]: If our component config reduces to a config map, should all other params shuffle to config map and be removed from the component config?  

Also we need a means to generate a config map with all the defaults.

Comment (timothysc) [inline pkg/apis/componentconfig/types.go]: Also do we need to know it's a configmap?  Isn't it just loading a file and the config map is an implementation detail.

**Comment by timothysc** [comment]
This is super important we get this right, and @k8s-mirror-cluster-lifecycle-misc will need to be involved.

**Review by timothysc — CHANGES_REQUESTED** [review]
I have some chicken-egg questions that need to be resolved 1st.

File: pkg/apis/componentconfig/types.go
Comment (bsalamat) [inline pkg/apis/componentconfig/types.go]: @timothysc This PR is not going to make any changes to scheduler's component config. This is only about scheduler's policy config. Policy config used to be a file whose path is provided in the component config. With this PR we support the file and also a ConfigMap to configure scheduler's policy.
This PR implements part of "Approach 2" as described in:
https://docs.google.com/document/d/19AKH6V6ejOeIvyGtIPNvRMR4Yi_X8U3Q1zz2fgTNhvM

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: ditto

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: ditto

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: prefer just after "stop".

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: This verification can not make sure the config from ConfigMap was took place; default config can also pass this test, right?

Similar to the following test, whether return `err` can not make sure the config was updated as expected; I'd like suggest to find a way to check the data/config with the scheduler.

File: pkg/apis/componentconfig/types.go
Comment (k82cn) [inline pkg/apis/componentconfig/types.go]: s/with this name/with this name in "kube-system" namespace/

Comment (k82cn) [inline pkg/apis/componentconfig/types.go]: s/with this name/with this name in "kube-system" namespace/

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/configurator.go]: swap `len(policyConfigMap.Data)` and `sc.policyConfigMap`.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/configurator.go]: prefer to get data directly and remove the comments :).

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: prefer just after "stop:=..."

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: The verification can not check whether the config from ConfigMap was set, as the default config can also pass it. 

Similar comments to the following cases, the `err` from `app.CreateScheduler` can not make sure the config was set in `sched`; prefer to check `sched`'s data directly or only enable part of `predicates`/`priorities` to check the result.

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: ditto

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (k82cn) [inline pkg/apis/componentconfig/v1alpha1/types.go]: ditto

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/options/options.go]: Done

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: I am not sure what you mean by this. `sc.policyConfigMap` is the name of the ConfigMap object. Here I am checking that the `Data` member of the ConfigMap object has only one element in its map.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: The `key` can be any arbitrary string and we don't know it.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: Please see my previous answer.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/configurator.go]: I mean the log :). Current output maybe `ConfigMap 1 has config_map_name entries in its ...`

File: test/integration/scheduler/scheduler_test.go
Comment (bsalamat) [inline test/integration/scheduler/scheduler_test.go]: Is there any style guide or recommendation on this? Other tests in the file use this pattern and I think consistency is more important, unless there is a strong reason to use a different pattern.

File: test/integration/scheduler/scheduler_test.go
Comment (bsalamat) [inline test/integration/scheduler/scheduler_test.go]: Scheduler does not expose its internal configuration for testing. I could change things in scheduler to make it testable, but  but I am not changing any logic around applying the config to the scheduler in this PR. That's why I thought those tests may not belong to this PR.

Comment (bsalamat) [inline test/integration/scheduler/scheduler_test.go]: I guess this is the same point as above. I will try to add a test.

File: pkg/apis/componentconfig/types.go
Comment (bsalamat) [inline pkg/apis/componentconfig/types.go]: Done

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/options/options.go]: Done.

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/options/options.go]: Good point. Done.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: Oh, sorry :) let me fix it.

Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: Done. Good catch, BTW! :)

File: pkg/apis/componentconfig/v1alpha1/types.go
Comment (bsalamat) [inline pkg/apis/componentconfig/v1alpha1/types.go]: Done

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: Yes, similar point; used to close and re-open browser, the previous comments was "lost" but it's here when I submit new comments :(. Sorry for the confuse.

The sched's did not expose data to check; but I think we can only enable part of `predicates/priorities` to check, e.g. disable "PortFit" predicates and dispatch two pods with same port to the same host , the second one should NOT be `Unscheduable`. But I think you will get a better test than that :).

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/configurator.go]: oh, yes. never mind :).

File: test/integration/scheduler/scheduler_test.go
Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: IMO, the style is from golang, prefer to call "defer" close to where the object was build, e.g. for "sync.Mutex", we'll call

```
l.Lock()
defer l.Unlock()

// other codes.
```

For the consistency, I'd like to correct other lines in this file :).

Comment (k82cn) [inline test/integration/scheduler/scheduler_test.go]: BTW, which line do you mean "Other tests in the file use this pattern"? Here's an example about `defer`.

```
 56 func TestUnschedulableNodes(t *testing.T) {
 57     _, s := framework.RunAMaster(nil)
 58     defer s.Close()
 59
 60     ns := framework.CreateTestingNamespace("unschedulable-nodes", s, t)
 61     defer framework.DeleteTestingNamespace(ns, s, t)
```

File: test/integration/scheduler/scheduler_test.go
Comment (davidopp) [inline test/integration/scheduler/scheduler_test.go]: I think it's probably fine to do in a different PR, but I think as part of this overall feature it's important to test that the values read from the ConfigMap were actually applied by the scheduler. (I realize we had no useful testing for the file case previously.) I think you were mentioning the other day that the factory test could be adapted to do this? Or something simple like what @k82cn suggested in the next comment (just tweak one policy and see that it is respected in how a pod is scheduled).

**Review by davidopp — COMMENTED** [review]
I only skimmed it, looks good overall, just a few comments.

**Comment by wanghaoran1988** [comment]
@bsalamat I think maybe you should update the [rbac bootstrap policy ](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/auth/authorizer/rbac/bootstrappolicy/policy.go#L337-L356)for the scheduler as you involve new resource type to access.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (wanghaoran1988) [inline plugin/cmd/kube-scheduler/app/configurator.go]: Is it better to add to the [defaults.go](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/componentconfig/v1alpha1/defaults.go#L133) ?

Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: Good point. Done.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: No, we don't. If it is empty, we ignore it.

**Comment by bsalamat** [comment]
@wanghaoran1988 is the rbac permissions needed to let scheduler access its ConfigMap?

**Comment by wanghaoran1988** [comment]
@bsalamat I am not completely sure about that, I met that problem when I update the scheduler lock location in this #42961 , you can have a try to enable rbac in your cluster.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/configurator.go]: seems brittle.

**Comment by timothysc** [comment]
Also rbac policy will likely need to be updated.  Google does not currently run with RBAC on by default.

**Review by timothysc — APPROVED** [review]

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: Why do you think it is brittle? When one imports an existing policy from a config file to a ConfigMap, the `Data` section of the map will get only one element. If there are more than one elements in the map, we won't know which one to use.

Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/configurator.go]: If a person makes a honest mistake it doesn't make a best effort to interpret and log a warning.  I won't block on this, but it's worth noting.

Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: The fact that we don't make a best effort is intentional. IMO it is better to fail to initialize the scheduler and let the user know that their config has errors than to initialize with an unintended config and surprise the user with the behavior. The latter would be harder to debug. You know how many warnings are thrown in each module and how effective they are in helping debug large systems.

Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/configurator.go]: I kind of wrote my second comment in haste, I completely agree with failing fast.  

I think my initial aversion here is that this almost seems like a logic portion of a obj.deserialization f(n), which should have it's own checks in place.

Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: We are using an existing mechanism for deserializing both the contents of a ConfigMap entry and a legacy config file. Before calling deserialization we need to make this quick sanity check. I agree that the sanity check belongs to deserialization code, but since we are sharing the deserialization with config file, I thought it would be fine to perform the check here.

**Comment by bsalamat** [comment]
I think I finally succeeded in adding tests that verify the scheduler's internal config. In order to do so, I had to expose some of the scheduler's internal state. Given that golang does not have the concept of immutable getters, the getters added should be used carefully. I added comments to note that these getters should be used for testing only, but I wish golang supported immutable getters.

File: test/integration/scheduler/scheduler_test.go
Comment (davidopp) [inline test/integration/scheduler/scheduler_test.go]: Update the comment to reflect that you now also check that the config is processed correctly.

**Comment by davidopp** [comment]
Thanks for improving the test!

File: test/integration/scheduler/scheduler_test.go
Comment (bsalamat) [inline test/integration/scheduler/scheduler_test.go]: Done. Thanks!

**Comment by timothysc** [comment]
lgtm to me now, if someone else wants to once over.

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/options/options.go]: The default value should be "kube-system"

File: plugin/pkg/scheduler/algorithm/scheduler_interface.go
Comment (k82cn) [inline plugin/pkg/scheduler/algorithm/scheduler_interface.go]: d/a pointer to/
or
s/pointer/reference/

**Comment by k82cn** [comment]
LGTM, minor comments.

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (wanghaoran1988) [inline plugin/cmd/kube-scheduler/app/options/options.go]: Wait, we do not need that if we have that set in the defaults.go @k82cn

**Comment by wanghaoran1988** [comment]
@liggitt @jayunit100 I am not sure whether openshift will enable user to use configMap change the policy config, but I know that openshift are try to reuse the controller rolebindings, if so, we need update the bootstrap policy for scheduler controller, and I prefer we have an default policy-configmap if we enabled them by default.

**Comment by liggitt** [comment]
This is a very similar discussion to the one kube-dns had, which wanted to make use of dynamic config, and started by making the pod watch a config map. That was modified to allow reading config from a file instead, which allowed mounting the config map holding the config into the pod. This meant no API entanglement or permissions needed for purposes of reading the config, and no modifications to the dns pod whether it was reading from a static file or one mounted from a config map.

File: plugin/cmd/kube-scheduler/app/options/options.go
Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/options/options.go]: When the `default.go` will execute? AFAIK, the default will be set by apiserver; not sure when the `componentconfig` is update.

Comment (k82cn) [inline plugin/cmd/kube-scheduler/app/options/options.go]: En, it's set in `NewSchedulerServer` manually. Also prefer to the set the default value to `kube-system` so the help message `[default=kube-system]` will be included.

Anyway, not a block comments.

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (liggitt) [inline plugin/cmd/kube-scheduler/app/configurator.go]: the brittle part is that we are not looking for a specific key. this blocks our ability to add anything to this config map in a compatible way.

Comment (liggitt) [inline plugin/cmd/kube-scheduler/app/configurator.go]: the scheduler should either look for a fixed key or be able to be told which key to use

**Comment by smarterclayton** [comment]
I would expect these comments to be fixed - is there an issue for that?

File: plugin/cmd/kube-scheduler/app/configurator.go
Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: @liggitt When an existing policy file is imported by `kubectl`, the key will be the name of the file and the content of the file will go in the value. I wanted to make it easy for customers to import their existing configs and use it. If more keys are needed in the config, they can be added to the content which is the "value" of the first element.

Comment (liggitt) [inline plugin/cmd/kube-scheduler/app/configurator.go]: accepting a single arbitrary key is confusing... kubectl can set the key when importing an existing config file (`--from-file=config.json=/path/to/my-scheduler-config.yaml`), and we should direct the user to do that.

Comment (liggitt) [inline plugin/cmd/kube-scheduler/app/configurator.go]: for example, if I have a working scheduler pointing at a config map, I would *not* expect adding a second key to break the scheduler

Comment (bsalamat) [inline plugin/cmd/kube-scheduler/app/configurator.go]: I see. I didn't know that kubectl can set the key. I will send a PR to remove the restriction on having only one element in the map and will add the logic to look for a specific key.

File: test/integration/scheduler/scheduler_test.go
Comment (liggitt) [inline test/integration/scheduler/scheduler_test.go]: Seeing test failures on this

Comment (liggitt) [inline test/integration/scheduler/scheduler_test.go]: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39716/pull-kubernetes-unit/26428/

**Pull request merged**
